scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1500 @@
+    lowest discrepancy. Then you could just update the discrepancy with
+    each candidate using `_update_discrepancy`. This method is faster than
+    computing the discrepancy for a large number of candidates.
+
+    References
+    ----------
+    .. [1] Fang et al. Design and modeling for computer experiments,
+       Computer Science and Data Analysis Series, 2006.
+    .. [2] Zhou Y.-D. et al. Mixture discrepancy for quasi-random point sets
+       Journal of Complexity, 29 (3-4) , pp. 283-301, 2013.
+    .. [3] T. T. Warnock. Computational investigations of low discrepancy point
+       sets, Applications of Number Theory to Numerical
+       Analysis, Academic Press, pp. 319-343, 1972.
+
+    Examples
+    --------
+    Calculate the quality of the sample using the discrepancy:
+
+    >>> from scipy.stats import qmc
+    >>> space = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+    >>> l_bounds = [0.5, 0.5]
+    >>> u_bounds = [6.5, 6.5]
+    >>> space = qmc.scale(space, l_bounds, u_bounds, reverse=True)
+    >>> space
+    array([[0.08333333, 0.41666667],
+           [0.25      , 0.91666667],
+           [0.41666667, 0.25      ],
+           [0.58333333, 0.75      ],
+           [0.75      , 0.08333333],
+           [0.91666667, 0.58333333]])
+    >>> qmc.discrepancy(space)
+    0.008142039609053464
+
+    """
+    sample = np.asarray(sample)
+
+    n, d = sample.shape
+
+    if iterative:
+        n += 1
+
+    if method == 'CD':
+        abs_ = abs(sample - 0.5)
+        disc1 = np.sum(np.prod(1 + 0.5 * abs_ - 0.5 * abs_ ** 2, axis=1))
+
+        prod_arr = 1
+        for i in range(d):
+            s0 = sample[:, i]
+            prod_arr *= (1 +
+                         0.5 * abs(s0[:, None] - 0.5) + 0.5 * abs(s0 - 0.5) -
+                         0.5 * abs(s0[:, None] - s0))
+        disc2 = prod_arr.sum()
+
+        return ((13.0 / 12.0) ** d - 2.0 / n * disc1 +
+                1.0 / (n ** 2) * disc2)
+    elif method == 'WD':
+        prod_arr = 1
+        for i in range(d):
+            s0 = sample[:, i]
+            x_kikj = abs(s0[:, None] - s0)
+            prod_arr *= 3.0 / 2.0 - x_kikj + x_kikj ** 2
+
+        return - (4.0 / 3.0) ** d + 1.0 / (n ** 2) * prod_arr.sum()
+    elif method == 'MD':
+        abs_ = abs(sample - 0.5)
+        disc1 = np.sum(np.prod(5.0 / 3.0 - 0.25 * abs_ - 0.25 * abs_ ** 2,
+                               axis=1))
+
+        prod_arr = 1
+        for i in range(d):
+            s0 = sample[:, i]
+            prod_arr *= (15.0 / 8.0 -
+                         0.25 * abs(s0[:, None] - 0.5) - 0.25 * abs(s0 - 0.5) -
+                         3.0 / 4.0 * abs(s0[:, None] - s0) +
+                         0.5 * abs(s0[:, None] - s0) ** 2)
+        disc2 = prod_arr.sum()
+
+        disc = (19.0 / 12.0) ** d
+        disc1 = 2.0 / n * disc1
+        disc2 = 1.0 / (n ** 2) * disc2
+
+        return disc - disc1 + disc2
+    elif method == 'star':
```

Comment:
What can I check this against? In [2], it is defined in a way that doesn't immediately suggest this algorithm.

Reply:
Yes I did not do anything with this one and just translated the maths. I will take your code then if you say it is the same.

Reply:
Oops, I didn't mean that my code should be used. It is very inefficient in terms of memory. I suggested that the difference in performance between them needed to be investigated more closely. I didn't take a close look, but I think the original code might have been doing some sort of work each loop unnecessarily, because the execution time seemed to be growing faster than it should. If we could fix that, we could make it fast _and_ more memory efficient.

I included these functions in the review for a few reasons:

- It is evidence that I carefully checked the accuracy of the `discrepancy` function
- They could be used in the test suite as an independent accuracy check for arbitrary data if we are unsatisfied with what we find in textbooks. Not only were they written independently - and based on a different source, it turns out - but they (well, the first three) are very easy to compare against the equations.

Reply:
Ok so I will take your experiments as a secondary implementation and test against that.

Reply:
For now I kept your version for `L2-star` as it is way faster than what I had. And I used mine for the test.